src/base/baseBrushTool.js

```diff
@@ -0,0 +1,367 @@
+  }
+
+  //===================================================================
+  // Implementation interface
+  //===================================================================
+
+  /**
+  * Event handler for KEY_DOWN event.
+  *
+  * @event
+  * @param {Object} evt - The event.
+  */
+  onKeyDown (evt) {
+    const eventData = evt.detail;
+    const element = eventData.element;
+    const keyCode = eventData.keyCode;
+
+    if (!isToolActive(element, this.name) || !this._keyboardController) {
+      return;
+    }
+
+    const imageNeedsUpdate = this._keyboardController.keyPress(keyCode);
+
+    if (imageNeedsUpdate) {
+      external.cornerstone.updateImage(element);
+    }
+  }
+
+  /**
+   * Draws the ImageBitmap the canvas.
+   *
+   * @protected
+   * @param  {Object} evt description
+   */
+  _drawImageBitmap (evt) {
+    const configuration = this.configuration;
+    const eventData = evt.detail;
+    const context = getNewContext(eventData.canvasContext.canvas);
+
+    const canvasTopLeft = external.cornerstone.pixelToCanvas(eventData.element, {
+      x: 0,
+      y: 0
+    });
+    const canvasBottomRight = external.cornerstone.pixelToCanvas(eventData.element, {
+      x: eventData.image.width,
+      y: eventData.image.height
+    });
+    const canvasWidth = canvasBottomRight.x - canvasTopLeft.x;
+    const canvasHeight = canvasBottomRight.y - canvasTopLeft.y;
+
+    context.imageSmoothingEnabled = false;
+    context.globalAlpha = configuration.brushAlpha;
+    context.drawImage(this._imageBitmap, canvasTopLeft.x, canvasTopLeft.y, canvasWidth, canvasHeight);
+    context.globalAlpha = 1.0;
+  }
+
+  /**
+   *  Changes the draw color (segmentation) of the tool.
+   *
+   * @protected
+   * @param  {Number} drawId The id of the color (segmentation) to switch to.
+   */
+  _changeDrawColor (drawId) {
+    const configuration = this._configuration;
+    const colormap = external.cornerstone.colors.getColormap(COLOR_MAP_ID);
+
+    configuration.draw = drawId;
+    const colorArray = colormap.getColor(configuration.draw);
+
+    configuration.hoverColor = `rgba(${colorArray[[0]]}, ${colorArray[[1]]}, ${colorArray[[2]]}, 0.8 )`;
+    configuration.dragColor = `rgba(${colorArray[[0]]}, ${colorArray[[1]]}, ${colorArray[[2]]}, 1.0 )`;
+  }
+
+  /**
+  * Event handler for MOUSE_UP during the drawing event loop.
+  *
+  * @protected
+  * @event
+  * @param {Object} evt - The event.
+  */
+  _drawingMouseUpCallback(evt) {
+    const eventData = evt.detail;
+    const element = eventData.element;
+
+    this.configuration.drawing = false;
+    this.configuration.mouseUpRender = true;
+
+    this._stopListeningForMouseUp(element);
+  }
+  /**
+  * Adds modify loop event listeners.
+  *
+  * @protected
+  * @param {Object} element - The viewport element to add event listeners to.
+  * @modifies {element}
+  */
+  _startListeningForMouseUp (element) {
+    element.removeEventListener(EVENTS.MOUSE_UP, this._drawingMouseUpCallback);
```

Comment:
Makes sense. We're locking ourselves to this tool while drawing, and this works like the handles in that an up/click will end the loop/lock?

Reply:
Yup :).